### PR TITLE
Enable `clippy` checks for yjit in CI

### PR DESCRIPTION
**Pull request by ianks**
This PR adds a job to check for [`clippy`](https://doc.rust-lang.org/stable/clippy/index.html).

> A collection of lints to catch common mistakes and improve your [Rust](https://github.com/rust-lang/rust) code.

The benefit of adding this as a CI step is that it can help ensure all code on the master branch is problem-free, and your text editor won't list problems for code you didn't write. Overall, I find this makes for a more welcoming development environment for new contributors, so is worth the maintenance cost.

99% of the code already adheres to the recommended lints. I just had to make a couple of small changes to make things fully compliant for CI.

### CODE REVIEW

**Comment by noahgibbs** [comment]
@maximecb - you've had strong feelings on this in the past. WDYT?

**Comment by jimmyhmiller** [comment]
I love clippy personally. And am always happy for changes that follow clippy's advice. What I personally don't want is failing CI because of clippy warnings. For me, CI tells us if the code compiles and works. Happy for something to remind us about some clippy things, like leaving a comment, but not failing CI.

**Comment by jimmyhmiller** [comment]
The CI failures here are because Cirrus is using the latest clippy. Once I upgraded locally, I got the same errors it got.

**Comment by ianks** [comment]
> For me, CI tells us if the code compiles and works. Happy for something to remind us about some clippy things, like leaving a comment, but not failing CI.

Ok, I can remove the `-D warnings` part for now so that it won't exit 0 when warnings are found. This won't leave a comment for now, but potentially something that can be added later.

**Comment by maximecb** [comment]
I'd rather people run clippy locally if they want to...

Otherwise, we could enable clippy on CI, but the CI should not fail because of it.

**Comment by ianks** [comment]
> I'd rather people run clippy locally if they want to...
> 
> Otherwise, we could enable clippy on CI, but the CI should not fail because of it.

Makes sense. Would my changes in https://github.com/ruby/ruby/pull/7093/commits/3d0fdce8b27a5de1fb36a8703ac53778abb55deb be OK then?

**Comment by maximecb** [comment]
> > I'd rather people run clippy locally if they want to...
> > Otherwise, we could enable clippy on CI, but the CI should not fail because of it.
> 
> Makes sense. Would my changes in [3d0fdce](https://github.com/ruby/ruby/commit/3d0fdce8b27a5de1fb36a8703ac53778abb55deb) be OK then?

Sure.

Curious why the CI is failing. Seems unrelated?

File: .github/workflows/yjit-ubuntu.yml [hunk 1]
```diff
@@ -40,6 +40,15 @@ jobs:
...
         working-directory: yjit
+  lint:
+    name: Rust lint
+    # GitHub Action's image seems to already contain a Rust 1.58.0.
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: actions/checkout@755da8c3cf115ac066823e79a1e1788f8940201b # v3.2.0
+      # Check that we don't have linting errors in release mode, too
+      - run: cargo clippy --all-targets --all-features
+        working-directory: yjit
```
Comment (maximecb) [inline .github/workflows/yjit-ubuntu.yml:51]: Will this also not fail if there are warnings?

**Comment by ianks** [comment]
> Curious why the CI is failing. Seems unrelated?

Yes looks unrelated. I just ran again so hopefully it self-heals.

**Review by maximecb — APPROVED** [review]

**Pull request merged**